runhouse/servers/http/http_server.py

```diff
@@ -1,43 +1,184 @@
 import traceback
+from functools import wraps
 from pathlib import Path
-from typing import Optional
+from typing import Optional, Union
 
 import ray
 import requests
-from fastapi import Body, FastAPI
+from fastapi import Body, FastAPI, HTTPException, Request
 from fastapi.encoders import jsonable_encoder
 from fastapi.responses import JSONResponse, StreamingResponse
 
 from sky.skylet.autostop_lib import set_last_active_time_to_now
 
 from runhouse.globals import configs, env_servlets, rns_client
+from runhouse.resources.hardware.utils import (
+    _current_cluster,
+    _get_cluster_from,
+    _load_cluster_config,
+)
+from runhouse.rns.utils.api import (
+    load_resp_content,
+    resolve_absolute_path,
+    ResourceAccess,
+)
 from runhouse.rns.utils.names import _generate_default_name
-from runhouse.servers.servlet import EnvServlet
-from ..http.http_utils import (
+from runhouse.servers.http.http_utils import (
     b64_unpickle,
-    DEFAULT_SERVER_HOST,
-    DEFAULT_SERVER_PORT,
     Message,
     OutputType,
     pickle_b64,
     Response,
+    ServerCache,
+    TLSCertConfig,
 )
+from runhouse.servers.nginx.config import NginxConfig
+from runhouse.servers.servlet import EnvServlet
 
 logger = logging.getLogger(__name__)
 
 app = FastAPI()
 
+# **NOTE**: This is a temp in-mem cache, we will move this out into the object store for future permissions support
+server_cache = ServerCache()
+
+global den_auth
+
+
+def validate_user(func):
+    """If using Den auth, validate the user's Runhouse token and access to the resource before continuing."""
+
+    @wraps(func)
+    async def wrapper(*args, **kwargs):
+        request: Request = kwargs.get("request")
+        func_call: bool = func.__name__ in ["call_module_method", "call"]
+        use_den_auth: bool = den_auth
+        is_coro = inspect.iscoroutinefunction(func)
+
+        if not use_den_auth:
+            if is_coro:
+                return await func(*args, **kwargs)
+
+            return func(*args, **kwargs)
+
+        token = request.headers.get("Authorization", "").split("Bearer ")[-1]
+        if not token:
+            raise HTTPException(
+                status_code=404,
+                detail="No token found in request auth headers. Expected in "
+                f"format: {json.dumps({'Authorization': 'Bearer <token>'})}",
+            )
+
+        # Check if user's token has already been validated and saved to cache on the cluster
+        cached_resources: dict = server_cache.get_resources(token)
+        if not cached_resources:
+            resp = requests.get(
+                f"{rns_client.api_server_url}/user",
+                headers={"Authorization": f"Bearer {token}"},
+            )
+            if resp.status_code != 200:
+                raise HTTPException(
+                    status_code=403,
+                    detail=f"Failed to validate Runhouse user: {load_resp_content(resp)}",
+                )
+
+        if use_den_auth or func_call:
```

Comment:
This should just be `if func_call`, if use_den_auth is false we return above

Reply:
And I don't see how we're validating auth into the cluster for the non-func_call case